bin/propolis-server/src/lib/stats/mod.rs

```diff
@@ -78,15 +82,80 @@ struct ServerStatsInner {
     /// data.
     virtual_machine: VirtualMachine,
 
+    /// A weak reference to the objects implementing the VM this Propolis server
+    /// is responsible for. It is critical this is a weak reference: we do not
+    /// want to prohibit `VmObjects` from being dropped when unneeded. The
+    /// referece is, however, necessary to measure "Propolis overhead" -
+    /// measurements of the Propolis process minus the VM it is managing.
+    vm_objects: Weak<VmObjects>,
+
     /// The reset count for the relevant instance.
     run_count: virtual_machine::Reset,
+
+    /// Mapped virtual memory for the Propolis managing this instance, excluding
+    /// VM memory.
+    vmm_vsz: virtual_machine::VmmVsz,
+
+    /// Process RSS is sampled in a standalone task. The `tokio::sync::watch`
+    /// here is to observe that sampling and update the `vmm_vsz` tracked
+    /// here.
+    process_stats_rx: tokio::sync::watch::Receiver<process::ProcessStats>,
 }
 
 impl ServerStatsInner {
-    pub fn new(virtual_machine: VirtualMachine) -> Self {
+    pub fn new(
+        virtual_machine: VirtualMachine,
+        vm_objects: Weak<VmObjects>,
+    ) -> Self {
+        let rx = process::ProcessStats::new();
+
         ServerStatsInner {
             virtual_machine,
+            vm_objects,
             run_count: virtual_machine::Reset { datum: Default::default() },
+            vmm_vsz: virtual_machine::VmmVsz { datum: Default::default() },
+            process_stats_rx: rx,
+        }
+    }
+
+    fn refresh_vmm_vsz(&mut self) {
+        let last_process_stats = self.process_stats_rx.borrow();
+
+        if let Some(vm_objects) = self.vm_objects.upgrade() {
+            // The mutex guarding VmObjects is async. Since we're not in an
+            // async context, we have to be a bit more explicit with the future.
+            // It's possible both for us to block acquiring this lock (if the
+            // state driver had an exclusive lock to make some change) and for
+            // us to block the state driver by acquiring the lock (but we
+            // *should* drop the lock very quickly, here).
```

Comment:
Yuck...I kinda wonder if it's worth doing this virtual address space accounting by having an atomic integer that we add to when mapping ranges, and which we can read from here without having to lock the whole `VmObjects` and iterate over every mapped region every time? That way, we go from "a critical section that's _O_(_n_mapped_physical_mem_regions_)" to "not having a critical section at all...

Reply:
ON FURTHER CONSIDERATION: I have been reminded that the lock around `VmObjects` is a `RwLock` rather than a `Mutex`, so we're not contending with other readers, and writes only occur on VM initialization/rundown. So...this probably isn't worth the effort.